examples/cmd/snapshotting/example_demo.go

```diff
@@ -0,0 +1,393 @@
+	const cniCacheDir = "/var/lib/cni"
+	const networkName = "fcnet"
+	const ifName = "veth0"
+
+	networkMask := "/24"
+	subnet := "10.168.0.0" + networkMask
+	var ipToRestore string
+
+	var maxRetries int = 10
+	var backoffTimeMs time.Duration = 500
+
+	sshKeyPath := filepath.Join(dir, "root-drive-ssh-key")
+
+	// Setup socket and snapshot + memory paths
+
+	socketPath := filepath.Join(dir, "snapshotssh")
+
+	err = os.Mkdir(socketPath, 0777)
+	if err != nil && !errors.Is(err, os.ErrExist) {
+		log.Fatal(err)
+	}
+
+	snapPath := filepath.Join(socketPath, "SnapFile")
+	memPath := filepath.Join(socketPath, "MemFile")
+
+	ctx := context.Background()
+
+	var createSnapshotSSH = func() {
+		// Network config
+		cniConfPath := fmt.Sprintf("%s/%s.conflist", cniConfDir, networkName)
+		err := writeCNIConfWithHostLocalSubnet(cniConfPath, networkName, subnet)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer os.Remove(cniConfPath)
+
+		networkInterface := sdk.NetworkInterface{
+			CNIConfiguration: &sdk.CNIConfiguration{
+				NetworkName: networkName,
+				IfName:      ifName,
+				ConfDir:     cniConfDir,
+				BinPath:     cniBinPath,
+				VMIfName:    "eth0",
+			},
+		}
+
+		cfg := createNewConfig(fmt.Sprintf("%s.create", socketPath),
+			withNetworkInterface(networkInterface),
+		)
+
+		// Use firecracker binary when making machine
+		cmd := sdk.VMCommandBuilder{}.WithSocketPath(fmt.Sprintf("%s.create", socketPath)).WithBin(filepath.Join(dir, "firecracker")).Build(ctx)
+
+		m, err := sdk.NewMachine(ctx, cfg, sdk.WithProcessRunner(cmd))
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer os.Remove(socketPath + ".create")
+
+		err = m.Start(ctx)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer func() {
+			if err := m.StopVMM(); err != nil {
+				log.Fatal(err)
+			}
+		}()
+		defer func() {
+			if err := m.Shutdown(ctx); err != nil {
+				log.Fatal(err)
+			}
+		}()
+
+		ipToRestore = m.Cfg.NetworkInterfaces[0].StaticConfiguration.IPConfiguration.IPAddr.IP.String()
+		fmt.Printf("IP of VM: %v\n", ipToRestore)
+		// fmt.Scanln()
+
+		var client *ssh.Client
+		for i := 0; i < maxRetries; i++ {
+			client, err = connectToVM(m, sshKeyPath)
+			if err != nil {
+				time.Sleep(backoffTimeMs * time.Millisecond)
+			} else {
+				break
+			}
+		}
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer client.Close()
+
+		session, err := client.NewSession()
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer session.Close()
+
+		fmt.Println(`Sending "sleep 422" command...`)
+		err = session.Start(`sleep 422`)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		fmt.Println("Creating snapshot...")
+
+		err = m.PauseVM(ctx)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		err = m.CreateSnapshot(ctx, memPath, snapPath)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		err = m.ResumeVM(ctx)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		fmt.Println("Snapshot created")
+	}
+
+	var loadSnapshotSSH = func() {
```

Comment:
just from glancing at this, I can review more next week - for this and the create snapshot function, is there a reason you chose to define these in `main` rather than as their own functions? I'm guessing maybe to keep some variables in scope?

Reply:
Made the changes and refactored some to make the functions separate, this ended up with me creating a global variable but I think that should be fine since it's just example code.